### ISSUE TITLE
Content not saved correctly when switching visual / code editor modes via keyboard shortcuts

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#24054]
**Describe the bug**

When switching Visual editor / Code editor modes and saving content via keyboard shortcuts, changes made to the content in ~edit~ Code editor mode aren't saved correctly.

A bit weirdly, content is saved correctly when using the controls provided the user interface instead of the keyboard shortcuts.

**To reproduce**

- WordPress 5.5 Beta 2
- create a post and add 3 paragraphs
- I tested with some very simple content:

```
<!-- wp:paragraph -->
<p>First paragraph.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Second paragraph.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Third paragraph.</p>
<!-- /wp:paragraph -->
```

- switch to Code editor via the keyboard shortcuts, on macOS: Shift+Alt+Cmd+M
- manually edit the third paragraph text and change it to "Hello"
- save as draft by pressing Cmd+S
- the saving happens with no errors:
  - "Save draft" in the toolbar changes to "Saved"
  - no errors in the console
- switch back to Visual editor by pressing Shift+Alt+Cmd+M
- observe the third paragraph content is still "Third paragraph."
- expected: it should be "Hello"

Repeat the steps above but this time, instead of using the keyboard shortcuts, use the controls in the UI:
- switch mode by clicking "Code editor" in the "More tools & option" menu
- edit the last paragraph content
- save by clicking "Save draft"
- switch back by clicking "Visual editor" in the "More tools & option" menu
- observe your edit has been saved correctly

**Comment by talldan** [comment]
I did some debugging and noticed this a couple of things. First, the post content in the request to the REST API is incorrect when following the reproduction steps, which explains why the changes in Code Editor aren't reflected.

The second thing I discovered is best illustrated by following these steps:
1. Switch to code editor
2. Make some changes
3. Click somewhere outside of the code editor text area
4. Use cmd+s to save
5. Switch back to visual mode
6. Observe content was updated correctly.

From this, it looks like the issue is that the in-memory state representing post content isn't being updated until too late. Seems to be related to the `stopEditing` call here which only happens on blur of the text area:
https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/post-text-editor/index.js#L82

This component hasn't changed in a long time, so I wonder if it's always been an issue or whether upstream changes in the state handling might have caused the issue 😓 

I'm poking around to see what a good fix would look like, but open to suggestions.

**Comment by afercia** [comment]
Thanks for looking at this and for the PR!
